engine/runtime/src/main/java/org/enso/interpreter/runtime/warning/AppendWarningNode.java

```diff
@@ -94,12 +96,58 @@ WithWarnings doMultipleWarningsArray(
     return new WithWarnings(value, warnsLimit, isLimitReached, warnsMap);
   }
 
+  /**
+   * This specialization should be the most frequent - just wrapping the given {@code object} with
+   * warnings hash map
+   */
+  @Specialization(guards = {"!isWithWarns(object)"})
+  WithWarnings doObjectMultipleWarningsHashMap(
+      VirtualFrame frame, Object object, EnsoHashMap newWarnsMap) {
+    assert !(object instanceof WithWarnings);
+    int warnsLimit = EnsoContext.get(this).getWarningsLimit();
+    return new WithWarnings(
+        object, warnsLimit, newWarnsMap.getHashSize() >= warnsLimit, newWarnsMap);
+  }
+
+  @Specialization
+  WithWarnings doWithWarnMultipleWarningsHashMap(
+      VirtualFrame frame,
+      WithWarnings withWarnings,
+      EnsoHashMap newWarnsMap,
+      @Shared @CachedLibrary(limit = "3") InteropLibrary interop,
+      @Exclusive @Cached ConditionProfile hasCachedVecReprProfile,
+      @Shared @Cached HashMapInsertNode mapInsertNode) {
+    if (withWarnings.isLimitReached()) {
+      return withWarnings;
+    }
+    var maxWarns = withWarnings.maxWarnings;
+    var newWarnsEntriesVec = newWarnsMap.getCachedVectorRepresentation(hasCachedVecReprProfile);
```

Comment:
I expect the `getCachedVectorRepresentation` to be quite slow operation. Given one just iterates thru it, we might rather use _iterator_ like concept.

Reply:
Using `HashMapInsertAllNode` in c2815e8a5d1064c2bbce1d8631b8c0da9d4e9d07. But I like the *iterator* idea.